pythonforandroid/pythonpackage.py

```diff
@@ -0,0 +1,664 @@
+                    {"home", ".tox"}.intersection(set(p.split(sep))) or \
+                    "users" in p.lower():
+                # Doesn't look like bog-standard system path.
+                if (p.endswith(os.path.sep + "bin") or
+                        p.endswith(os.path.sep + "bin" + os.path.sep)):
+                    # Also ends in "bin" -> likely virtualenv/venv.
+                    # Add as unfavorable / end of candidates:
+                    bad_candidates.append(p)
+                    continue
+            ever_had_nonvenv_path = True
+
+        good_candidates.append(p)
+
+    # See if we can now actually find the system python:
+    for p in good_candidates + bad_candidates:
+        result = python_binary_from_folder(p)
+        if result is not None:
+            return result
+
+    raise RuntimeError("failed to locate system python in: " +
+                       sys.real_prefix)
+
+
+def get_package_as_folder(dependency, major_python_version=3):
+    """ This function downloads the given package / dependency and extracts
+        the raw contents into a folder.
+
+        Afterwards, it returns a tuple with the type of distribution obtained,
+        and the temporary folder it extracted to.
+
+        Examples of returned values:
+
+        ("source", "/tmp/pythonpackage-venv-e84toiwjw")
+        ("wheel", "/tmp/pythonpackage-venv-85u78uj")
+
+        What the distribution type will be depends on what pip decides to
+        download.
+    """
+
+    venv_parent = tempfile.mkdtemp(
+        prefix="pythonpackage-venv-"
+    )
+    try:
+        # Create a venv to install into:
+        try:
+            if int(sys.version.partition(".")[0]) < 3:
+                # Python 2.x has no venv.
+                subprocess.check_output([
+                    _get_system_python_executable(),
+                    "-m", "virtualenv",
+                    "--python=" + _get_system_python_executable(),
+                    os.path.join(venv_parent, 'venv')
+                ], cwd=venv_parent)
+            else:
+                # On modern Python 3, use venv.
+                subprocess.check_output([
+                    _get_system_python_executable(), "-m", "venv",
+                    os.path.join(venv_parent, 'venv')
+                ], cwd=venv_parent)
+        except subprocess.CalledProcessError as e:
+            output = e.output
+            try:
+                output = output.decode('utf-8', 'replace')
+            except AttributeError:
+                pass
+            raise ValueError(
+                'venv creation unexpectedly ' +
+                'failed. error output: ' + str(output)
+            )
+        venv_path = os.path.join(venv_parent, "venv")
+
+        # Update pip and wheel in venv for latest feature support:
+        try:
+            filenotfounderror = FileNotFoundError
+        except NameError:  # Python 2.
+            filenotfounderror = OSError
```

Comment:
ouch, I think `six` is already in our deps, maybe we should use it here

Reply:
not sure? I actually experimented around a bit and tried to assign FileNotFoundError directly on python 2, but I couldn't get it to work. (it would lead to an odd NameError, I must have confused the scoping somehow by messing with the built-in)

will resolve for now since you haven't shared a good idea and I agree it's ugly but I don't have another either and it does seem fairly minor. but feel free to poke/slap and/or unresolve if you come up with something! :+1: